### PR TITLE
[NFC] Refactor Overload Resolution a Bit

### PR DESCRIPTION
**Pull request by CodaFi**
The first commit pulls the out-parameters in `resolveOverloadForDeclWithSpecialTypeCheckingSemantics` out as its return type and adjusts the caller in `ConstraintSystem::resolveOverload` to have a bit of a simpler structure. 

The second commit further clarifies the structure of `ConstraintSystem::resolveOverload` by pulling out the parts that readjust the constraint system and the type of the reference to the selected overload into their own private helper function.  It also readjusts its structure from a branch-tree on the overload kind to a switch.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Review by xedin — APPROVED** [review]

File: lib/Sema/ConstraintSystem.cpp [hunk 1]
```diff
@@ -1587,21 +1587,16 @@ void ConstraintSystem::addOverloadSet(ArrayRef<Constraint *> choices,
...
-                                                     OverloadChoice choice,
-                                                     Type &refType,
-                                                     Type &openedFullType) {
-  assert(choice.getKind() == OverloadChoiceKind::Decl);
-
-  switch (TypeChecker::getDeclTypeCheckingSemantics(choice.getDecl())) {
+/// checking semantics, compute the type of the reference.  For now, follow
+/// the lead of \c getTypeOfMemberReference and return a pair of
+/// the full opened type and the reference's type.
+static std::pair<Type, Type> getTypeOfReferenceWithSpecialTypeCheckingSemantics(
```
Comment (slavapestov) [inline lib/Sema/ConstraintSystem.cpp]: It looks like both elements of the pair are equal in every case below? Should this just return a single Type?

Comment (CodaFi) [reply lib/Sema/ConstraintSystem.cpp]: It should, but I'm trying to maintain some consistency here while I go clean up the rest of the callers. I want to see if I can remove *all* the pairs.

Comment (CodaFi) [reply lib/Sema/ConstraintSystem.cpp]: *Consistency with respect to the `std::tie`s in the caller, I should say.

**Comment by CodaFi** [comment]
@swift-ci please smoke test Linux platform

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**